File: sdk/elixir/lib/telecom_sdk.ex
defmodule TelecomSDK do
  @moduledoc """
  Telecom Platform Elixir SDK

  Provides GenServer-based async access to the Telecom Platform API with full Elixir pattern matching.
  """

  use GenServer
  require Logger

  # Client API

  @doc """
  Start the Telecom SDK GenServer
  """
  def start_link(opts \\ []) do
    config = struct(Config, opts)
    GenServer.start_link(__MODULE__, config, name: __MODULE__)
  end

  @doc """
  Generate a JWT token for authentication
  """
  def generate_jwt_token(user_id, expiry_hours \\ 24, additional_claims \\ %{}) do
    GenServer.call(__MODULE__, {:generate_jwt_token, user_id, expiry_hours, additional_claims})
  end

  @doc """
  Validate a JWT token
  """
  def validate_jwt_token(token) do
    GenServer.call(__MODULE__, {:validate_jwt_token, token})
  end

  @doc """
  Close the SDK and cleanup resources
  """
  def close do
    GenServer.call(__MODULE__, :close)
  end

  # GenServer Callbacks

  @impl true
  def init(config) do
    auth_provider = AuthProvider.new(api_key: config.api_key, jwt_secret: config.jwt_secret)
    http_client = HTTPClient.new(config.api_url, auth_provider,
      timeout: config.timeout,
      max_retries: config.max_retries,
      retry_delay: config.retry_delay,
      enable_logging: config.enable_logging
    )

    # Initialize API modules
    subscribers = SubscriberAPI.new(http_client)
    usage = UsageAPI.new(http_client)
    payments = PaymentAPI.new(http_client)
    rating_plans = RatingPlanAPI.new(http_client)
    system = SystemAPI.new(http_client)
    analytics = AnalyticsAPI.new(http_client)
    security = SecurityAPI.new(http_client)
    currency = CurrencyAPI.new(http_client)

    state = %{
      config: config,
      auth_provider: auth_provider,
      http_client: http_client,
      subscribers: subscribers,
      usage: usage,
      payments: payments,
      rating_plans: rating_plans,
      system: system,
      analytics: analytics,
      security: security,
      currency: currency
    }

    {:ok, state}
  end

  @impl true
  def handle_call({:generate_jwt_token, user_id, expiry_hours, additional_claims}, _from, state) do
    auth_provider = AuthProvider.generate_jwt_token(state.auth_provider, user_id, expiry_hours, additional_claims)
    {:reply, {:ok, auth_provider.token_cache}, %{state | auth_provider: auth_provider}}
  end

  @impl true
  def handle_call({:validate_jwt_token, token}, _from, state) do
    result = AuthProvider.validate_jwt_token(state.auth_provider, token)
    {:reply, {:ok, result}, state}
  end

  @impl true
  def handle_call(:close, _from, state) do
    HTTPClient.close(state.http_client)
    {:reply, :ok, state}
  end

  # Configuration

  defstruct [:api_url, :api_key, :jwt_secret, :timeout, :max_retries, :retry_delay, :enable_logging]

  defmodule Config do
    @moduledoc """
    Configuration for Telecom SDK
    """

    defstruct [
      api_url: "http://localhost:8000",
      api_key: nil,
      jwt_secret: nil,
      timeout: 30_000,
      max_retries: 3,
      retry_delay: 1_000,
      enable_logging: false
    ]
  end
end

File: sdk/kotlin/src/main/kotlin/com/telecom/TelecomSDK.kt
package com.telecom

/**
 * Telecom Platform Kotlin SDK
 * 
 * Provides coroutine-based access to the Telecom Platform API with full Kotlin serialization support.
 */
class TelecomSDK private constructor(
    private val config: TelecomConfig,
    private val authProvider: AuthProvider,
    private val httpClient: HTTPClient
) {
    
    // API modules
    val subscribers: SubscriberAPI
    val usage: UsageAPI
    val payments: PaymentAPI
    val ratingPlans: RatingPlanAPI
    val system: SystemAPI
    val analytics: AnalyticsAPI
    val security: SecurityAPI
    val currency: CurrencyAPI
    
    companion object {
        /**
         * Create a new TelecomSDK instance
         */
        fun create(config: TelecomConfig = TelecomConfig()): TelecomSDK {
            val authProvider = AuthProvider(apiKey = config.apiKey, jwtSecret = config.jwtSecret)
            val httpClient = HTTPClient(config = config, authProvider = authProvider)
            
            // Initialize API modules
            val subscribers = SubscriberAPI(httpClient)
            val usage = UsageAPI(httpClient)
            val payments = PaymentAPI(httpClient)
            val ratingPlans = RatingPlanAPI(httpClient)
            val system = SystemAPI(httpClient)
            val analytics = AnalyticsAPI(httpClient)
            val security = SecurityAPI(httpClient)
            val currency = CurrencyAPI(httpClient)
            
            return TelecomSDK(config, authProvider, httpClient, subscribers, usage, payments, ratingPlans, system, analytics, security, currency)
        }
    }
    
    private constructor(
        config: TelecomConfig,
        authProvider: AuthProvider,
        httpClient: HTTPClient,
        subscribers: SubscriberAPI,
        usage: UsageAPI,
        payments: PaymentAPI,
        ratingPlans: RatingPlanAPI,
        system: SystemAPI,
        analytics: AnalyticsAPI,
        security: SecurityAPI,
        currency: CurrencyAPI
    ) {
        this.config = config
        this.authProvider = authProvider
        this.httpClient = httpClient
        this.subscribers = subscribers
        this.usage = usage
        this.payments = payments
        this.ratingPlans = ratingPlans
        this.system = system
        this.analytics = analytics
        this.security = security
        this.currency = currency
    }
    
    // Authentication methods
    
    /**
     * Generate a JWT token for authentication
     */
    fun generateJWTToken(userId: String, expiryHours: Int, additionalClaims: Map<String, Any> = emptyMap()): String {
        return authProvider.generateJWTToken(userId, expiryHours, additionalClaims)
    }
    
    /**
     * Validate a JWT token
     */
    fun validateJWTToken(token: String): Map<String, Any> {
        return authProvider.validateJWTToken(token)
    }
    
    // Cleanup
    
    /**
     * Close the SDK and cleanup resources
     */
    fun close() {
        httpClient.close()
    }
}

/**
 * Configuration for Telecom SDK
 */
data class TelecomConfig(
    val apiURL: String = "http://localhost:8000",
    val apiKey: String? = null,
    val jwtSecret: String? = null,
    val timeout: java.time.Duration = java.time.Duration.ofSeconds(30),
    val maxRetries: Int = 3,
    val retryDelay: Long = 1 // in seconds
)
